### ISSUE TITLE
[clang:static analyzer]  CTU false positive: `core.NullDereference`

### ISSUE THREAD

**Issue by bharsesh** [llvm/llvm-project#206653]
clang Static Analyzer reports a false positive `core.NullDereference` under CTU analysis for this test case (See attachment).

dep = bar(e);
if (dep != NULL && e->a > 0)

In this program, `bar()` returns non-null only when `e != NULL`, so the dereference of `e->a` is infeasible on the `dep != NULL` branch.

The same code does not warn when `bar()` and `foo()` are placed into a single translation unit and analyzed with `clang --analyze`.

This looks like a CTU precision issue where the analyzer loses the implication:
- dep = bar(e)
- dep != NULL
- therefore e != NULL

Observed diagnostic:
main.c:12:24: warning: Access to field 'a' results in a dereference of a null pointer (loaded from variable 'e') [core.NullDereference]
   12 |     if (dep != NULL && e->a > 0)
      |                        ^~~~

The issue can be reproduced with clang version 23.0.0git 
[testcase.rtf](https://github.com/user-attachments/files/29493156/testcase.rtf)

**Comment by chrchr-github** [comment]
Reduced:
```c
struct S {
    int a;
};

struct S* bar(struct S* e);

void foo(struct S* e) {
    struct S* dep = nullptr;
    dep = bar(e);

    if (dep != nullptr && e->a > 0) {}
}

int main() {
    struct S* e_ptr = nullptr;
    foo(e_ptr);
}
```
https://godbolt.org/z/Msjjxf6K3

**Comment by bharsesh** [comment]
If bar() is only declared and its definition is unavailable to the analyzer, the warning is fair.

  In that case the analyzer only sees:
  dep = bar(e);
  if (dep != NULL && e->a > 0)

  and so it cannot soundly infer dep != NULL => e != NULL, because an arbitrary unknown bar()
  could return a non-null pointer even when e is null. So the reduced version is not the same bug;
  it is just ordinary conservative behavior with an unknown callee.
